exec/java-exec/src/test/java/org/apache/drill/exec/store/parquet/TestParquetMetadataCache.java

```diff
@@ -737,6 +738,7 @@ public void testBooleanPartitionPruning() throws Exception {
     }
   }
 
+  @Ignore
```

Comment:
Could you please explain the reason why these tests should be ignored?

Reply:
@vvysotskyi no worries, work is still in progress.

Reply:
In this case, I don't think it's a good idea to disable unit tests. You can consider asking for help to resolve unit tests failures but not disable them.

Reply:
The test needs to be fixed as part of a separate JIRA/PR (another option is to remove the check for the filter, but IMO it is even less desirable).

Reply:
At what point do these tests fail? If they happen (as I suspect) during planning or building the metadata cache, then we are looking at a bigger issue that may need to be addressed before we switch to the upgraded version.
Essentially, pruning checks stats to determine whether the file should be read or not. If the stats are corrupted and we use them anyway, the query will return wrong results. If the stats are not corrupted and the Parquet library refuses to read them because it cannot be sure of the state of the stats, then the pruning will fail and the query will take too long to run. 
If the tests are failing to execute because the Parquet library fails to read the stats, then we will have a situation where queries that used to run correctly will stop running after we upgrade to the new Parquet library, which is likely to be *very* unpopular.

Reply:
The tests fail during plan validation as the new version of the parquet library ignores wrong statistics for the data types used by queries. Even if statistics is wrong for a small portion of parquet files and for the parquet files used by the tests it is correct, Drill can't rely on wrong statistics as it leads to the wrong query results. Basically, there is a bug in the version of the parquet library that Drill currently uses that may cause query result to be wrong and this bug is fixed in the new version causing 2 unit tests failures. IMO, it is better to upgrade to the new library sooner than later even if it will cause slowdown for some queries.

Reply:
Hmm, we need to take a look at this. For a period of two years, files written by tools using Parquet libraries were writing incorrect statistics, but because Drill had its own build where we had fixed the issue (we found the issue in the first place), files written by Drill were correct. A very large number of Drill users use the Parquet files produced by Drill and it was decided that we cannot penalize them. We provided a migration tool to users to tag files produced by Drill. The tool added information in the extra metadata in Parquet files to indicate the file was written by Drill and stats from these files should be allowed. 
AFAIK, this should be in the current build of Drill Parquet as well as in the Parquet library v 1.8.2 and above. 
Do you know if the stats corruption that affects these tests is something that was fixed in a version after 1.8.2?

Reply:
It will be good if you can point to JIRA with the fix that Drill uses to correct statistics. Without JIRA it is not clear what particular fix is used by Drill to workaround bugs in how parquet library handles statistics and for what data types.

Reply:
I have investigated why these tests fail. For example, let's take `testIntervalDayPartitionPruning`.
First test creates partitioned table using Drill. Since table is created at runtime, new parquet lib is used. Created table contains 4 files, one of them contains all nulls.  For this file with nulls, statistics for all types except of binary is `num_nulls: 3, min/max not defined`. For binary type it is `no stats for this column`. In previous parquet version, statistics was written correctly. Maybe this is bug in parquet, maybe in Drill writer. 

Another problem is with metadata file. We do write metadata for binary columns into it successfully. Example:
```
  "columnTypeInfo" : {
    "`col_intrvl_day`" : {
      "name" : [ "col_intrvl_day" ],
      "primitiveType" : "FIXED_LEN_BYTE_ARRAY",
      "originalType" : "INTERVAL",
      "precision" : 0,
      "scale" : 0,
      "repetitionLevel" : 0,
      "definitionLevel" : 1
    },
        "name" : [ "col_intrvl_day" ],
        "minValue" : "AAAAABoAAACQ4KEB",
        "maxValue" : "AAAAABoAAACQ4KEB",
        "nulls" : 0
```
But when reading it back from file, we read empty strings. Looks like this one is Drill bug.

@vrozov  I also have noticed that `ParquetFileReader.readFooter(conf, path, NO_FILTER);`. If you'll have a chance, please replace it.

Reply:
Where the `columnTypeInfo` comes from? Is it an output of parquet-tools? If yes, please provide details of the command line parameters used. If not, please check what parquet-tools report.

Reply:
This is snippet from metadata file. Drill generates it using info from the footer.

https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/metadata/Metadata.java

https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/metadata/Metadata_V3.java

Reply:
Please check what parquet-tools report when it reads that file back.

Reply:
I attached the output of `org.apache.parquet.format.converter.ParquetMetadataConverter` in the debug mode to [DRILL-6353](https://issues.apache.org/jira/browse/DRILL-6353). As you can see there is no statistics available for the `col_intrvl_day`:
```
{
  "encodingStats" : null,
  "dictionaryPageOffset" : 0,
  "valueCount" : 6,
  "totalSize" : 92,
  "totalUncompressedSize" : 91,
  "statistics" : {
    "max" : null,
    "min" : null,
    "maxBytes" : null,
    "minBytes" : null,
    "empty" : true,
    "numNulls" : -1,
    "numNullsSet" : false
  },
  "firstDataPageOffset" : 532,
  "type" : "FIXED_LEN_BYTE_ARRAY",
  "path" : [ "col_intrvl_day" ],
  "primitiveType" : {
    "name" : "col_intrvl_day",
    "repetition" : "OPTIONAL",
    "originalType" : "INTERVAL",
    "id" : null,
    "primitive" : true,
    "primitiveTypeName" : "FIXED_LEN_BYTE_ARRAY",
    "decimalMetadata" : null,
    "typeLength" : 12
  },
  "codec" : "SNAPPY",
  "encodings" : [ "RLE", "BIT_PACKED", "PLAIN" ],
  "startingPos" : 532
}
```
This is result of parquet fix for "Deprecate type-defined sort ordering for INTERVAL type"([PARQUET-1064](https://issues.apache.org/jira/browse/PARQUET-1064))

Reply:
The same applies to `testIntervalYearPartitionPruning`: statistics for `col_intrvl_yr` is also not available for the same reason:
```
{
  "encodingStats" : null,
  "dictionaryPageOffset" : 0,
  "valueCount" : 6,
  "totalSize" : 81,
  "totalUncompressedSize" : 91,
  "statistics" : {
    "max" : null,
    "min" : null,
    "maxBytes" : null,
    "minBytes" : null,
    "empty" : true,
    "numNulls" : -1,
    "numNullsSet" : false
  },
  "firstDataPageOffset" : 451,
  "type" : "FIXED_LEN_BYTE_ARRAY",
  "path" : [ "col_intrvl_yr" ],
  "primitiveType" : {
    "name" : "col_intrvl_yr",
    "repetition" : "OPTIONAL",
    "originalType" : "INTERVAL",
    "id" : null,
    "primitive" : true,
    "primitiveTypeName" : "FIXED_LEN_BYTE_ARRAY",
    "decimalMetadata" : null,
    "typeLength" : 12
  },
  "codec" : "SNAPPY",
  "encodings" : [ "RLE", "BIT_PACKED", "PLAIN" ],
  "startingPos" : 451
}
```

Reply:
And for `testDecimalPartitionPruning` statistics for `MANAGER_ID` is not available either:
```
{
  "encodingStats" : null,
  "dictionaryPageOffset" : 0,
  "valueCount" : 107,
  "totalSize" : 168,
  "totalUncompressedSize" : 363,
  "statistics" : {
    "max" : null,
    "min" : null,
    "maxBytes" : null,
    "minBytes" : null,
    "empty" : true,
    "numNulls" : -1,
    "numNullsSet" : false
  },
  "firstDataPageOffset" : 5550,
  "type" : "FIXED_LEN_BYTE_ARRAY",
  "path" : [ "MANAGER_ID" ],
  "codec" : "SNAPPY",
  "primitiveType" : {
    "name" : "MANAGER_ID",
    "repetition" : "OPTIONAL",
    "originalType" : "DECIMAL",
    "id" : null,
    "primitive" : true,
    "primitiveTypeName" : "FIXED_LEN_BYTE_ARRAY",
    "decimalMetadata" : {
      "precision" : 6,
      "scale" : 0
    },
    "typeLength" : 3
  },
  "encodings" : [ "PLAIN", "BIT_PACKED", "RLE" ],
  "startingPos" : 5550
}
```

Reply:
Parquet library behavior for DECIMAL statistics was changed in [PARQUET-686](https://issues.apache.org/jira/browse/PARQUET-686) (see [Parquet PR #367](https://github.com/apache/parquet-mr/pull/367)). I filed [PARQUET-1322](https://issues.apache.org/jira/browse/PARQUET-1322) to track statistics availability for DECIMAL types.

Reply:
Vlad thanks for investigating the issue. Since it's Parquet problem, it can leave tests to be ignored just please add comment in each of them to indicated the root cause. @parthchandra are you ok with this approach?

Reply:
I also had an offline chat with Vlad on this one. The problem is that Parquet has changed its behaviour and will not give us the stats for Decimal when we read footers. 
We have, therefore, no way of knowing whether Decimal stats are correct or not (even if they are correct) unless we try to hack something in Parquet. Hacking something in Parquet is not an option since that is exactly what this PR is trying to fix !
Also, we have never supported Decimal in Drill, so we do not have to consider backward compatibility. There are some users using Decimal (based on posts to the mailing list), but the old implementation never worked reliably so this will be an overall improvement for all parties.

+1. And thanks Vlad, Arina for pursuing this one to the end :)